### PR TITLE
Fix temp path creation in Dune cache

### PR DESCRIPTION
**Pull request by snowleopard**
See https://github.com/ocaml/dune/issues/4167.

### CODE REVIEW

File: otherlibs/stdune-unstable/temp.ml [hunk 1]
```diff
@@ -103,3 +104,11 @@ let temp_path ~dir ~prefix ~suffix =
         raise Exit
       else
         candidate)
+
+let with_temp_path ~dir ~prefix ~suffix ~f =
+  match temp_path ~dir ~prefix ~suffix with
+  | exception e -> f (Error e)
+  | temp_path ->
+    let result = f (Ok temp_path) in
```
Comment (rgrinberg) [inline otherlibs/stdune-unstable/temp.ml:112]: This needs a call to protect to make sure that the temporary path is deleted even if f raises.

**Review by rgrinberg — CHANGES_REQUESTED** [review]

File: otherlibs/stdune-unstable/temp.ml (hunk 1 above)
Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:112]: Good point, will do!

**Comment by rgrinberg** [comment]
A CHAGNES entry would also be nice here. Since this fix is important, it would be nice to make a point release that includes it as well.

**Comment by snowleopard** [comment]
> A CHAGNES entry would also be nice here. Since this fix is important, it would be nice to make a point release that includes it as well.

@rgrinberg Agreed, done!

**Review by rgrinberg — APPROVED** [review]

**Comment by snowleopard** [comment]
@rgrinberg Apparently `try_times` does return a meaningful exception, so I printed it out too. (At first read, I thought we hit the final `assert` if all the attempts fail.)

Thanks for the review!

**Pull request merged**